multicluster/service-mirror/cluster_watcher.go

```diff
@@ -718,23 +718,72 @@ func (rcsw *RemoteClusterServiceWatcher) handleFederatedServiceJoin(ctx context.
+		isOldest = true
+	} else if _, localDiscoveryExists := ev.localService.Annotations[consts.LocalDiscoveryAnnotation]; localDiscoveryExists {
+		// The local cluster is older than us.
+		isOldest = false
+	} else {
+		// The local cluster is not a member of the federated service. Therefore,
+		// we check the remote discovery to see if we are the oldest.
+		isOldest = true
+		members := strings.Split(ev.localService.Annotations[consts.RemoteDiscoveryAnnotation], ",")
+		for _, member := range members {
+			target := strings.Split(member, "@")
+			if len(target) != 2 {
+				rcsw.log.Errorf("Invalid federated service member: %s", member)
+				continue
+			}
+			cluster := target[1]
+			link, err := rcsw.localAPIClient.Link().Lister().Links(rcsw.serviceMirrorNamespace).Get(cluster)
+			if err != nil {
+				rcsw.log.Errorf("Failed to get link %s: %s", cluster, err)
+				continue
+			}
+			if link.CreationTimestamp.Before(&rcsw.link.CreationTimestamp) {
+				rcsw.log.Debugf("ignoring metadata from %s because %s is older", rcsw.link.Spec.TargetClusterName, cluster)
+				isOldest = false
+				break
+			}
+		}
+	}
+
+	if isOldest {
+		preservedAnnotations := map[string]string{}
+		for _, k := range []string{consts.RemoteDiscoveryAnnotation, consts.LocalDiscoveryAnnotation} {
+			if v, ok := ev.localService.Annotations[k]; ok {
+				preservedAnnotations[k] = v
+			}
+		}
+		ev.localService.Labels = rcsw.getFederatedServiceLabels(ev.remoteUpdate)
+		ev.localService.Annotations = rcsw.getFederatedServiceAnnotations(ev.remoteUpdate)
+		for k, v := range preservedAnnotations {
+			ev.localService.Annotations[k] = v
+		}
+
+		ev.localService.Spec.Ports = remapRemoteServicePorts(ev.remoteUpdate.Spec.Ports)
+	}
+
 	if rcsw.link.Spec.TargetClusterName == "" {
 		// Local discovery
 		ev.localService.Annotations[consts.LocalDiscoveryAnnotation] = ev.remoteUpdate.Name
 	} else {
 		// Remote discovery
 		remoteTarget := fmt.Sprintf("%s@%s", ev.remoteUpdate.Name, rcsw.link.Spec.TargetClusterName)
-		if remoteDiscoveryContains(ev.localService.Annotations[consts.RemoteDiscoveryAnnotation], remoteTarget) {
-			return nil
-		}
-		if ev.localService.Annotations[consts.RemoteDiscoveryAnnotation] == "" {
-			ev.localService.Annotations[consts.RemoteDiscoveryAnnotation] = remoteTarget
-		} else {
-			ev.localService.Annotations[consts.RemoteDiscoveryAnnotation] = fmt.Sprintf(
-				"%s,%s",
-				ev.localService.Annotations[consts.RemoteDiscoveryAnnotation],
-				remoteTarget,
-			)
+		if !remoteDiscoveryContains(ev.localService.Annotations[consts.RemoteDiscoveryAnnotation], remoteTarget) {
+			if ev.localService.Annotations[consts.RemoteDiscoveryAnnotation] == "" {
+				ev.localService.Annotations[consts.RemoteDiscoveryAnnotation] = remoteTarget
+			} else {
+				ev.localService.Annotations[consts.RemoteDiscoveryAnnotation] = fmt.Sprintf(
+					"%s,%s",
+					ev.localService.Annotations[consts.RemoteDiscoveryAnnotation],
+					remoteTarget,
+				)
+			}
```

Comment:
We were bailing out when the local service already had accounted for the target cluster in the `multicluster.linkerd.io/remote-discovery` annotation. IIUC with this change we will trigger a service update (and link status update) every time even if there are no changes, which might raise performance concerns in large deployments. Any chance to avoid that?